### ISSUE TITLE
Create a better position testing strategy 

### ISSUE THREAD

**Issue by heshanpadmasiri** [ballerina-platform/nballerina#816]
We need a better position testing strategy (to replace `testPos.bal`). Current expectation are

1. Use a "position centric tree" (similar to `astString.bal`)
2. Instead of assertion failures we should provide error messages more easier to understand (similar to compile errors showing positions with the source code)

**Comment by jclark** [comment]
Position-centric tree is something like this

```ballerina
type SyntaxNode TerminalSyntaxNode|NonTerminalSyntaxNode;

type TerminalSyntaxNode record {|
  *PositionFields;
  string? token;  // if non-nil expect the token to be this
|};

type NonTerminalSyntaxNode record {|
  *PositionFields;
  SyntaxNode[] childNodes;
|};
```

You will probably need some more complexity to deal with cases where we do not have enough information to construct this.

Basic concept is
- construct the position-centric tree from the AST
- validate the position-centric tree

For a non-terminal validation means that the ranges of the child nodes cover (without overlap) the range of the parent node, except for whitespace gaps.

Shouldn't need to do reparsing.

There will be some complexity from not explicitly representing parenthesized expressions/type descriptors.

**Comment by jclark** [comment]
We could have nodes in the position-oriented tree referring to nodes in the AST by just typing the AST nodes as

```ballerina
type AstNode record {|
   *PositionFields;
   any...;
|};
```

Constructing this position-oriented tree will I think be a useful step towards editor support.  We could potentially use it to map from a position in the file to a position in the syntax tree.

We could also (in the future) redo the AST so that the nodes are subtypes of more generic SyntaxNodes as above.

**Comment by jclark** [comment]
Another complexity will be switching the tokenizer between nodes while doing verification.

**Comment by jclark** [comment]
Fixed by #843
